content/blog/2019-09-25-autograd.md

```diff
@@ -0,0 +1,297 @@
+
+### Reverse-Mode vs Forward-Mode
+
+It may seem that, in typical contexts, reverse-mode automatic differentiation
+is strictly superior. It's far more common to see a "many to one" function
+than a "one to many" function. And that is true! Most automatic
+differentiation frameworks (i.e: ML frameworks) only implement reverse-mode
+(PyTorch, TensorFlow, MxNet, etc.).
+
+However, there are still instances where forward-mode automatic
+differentiation is very useful. For example, computing the Hessian of `f(x)`
+multiplied by a vector `v` - a Hessian vector product. The typical algorithm
+for this requires [both forward-mode and reverse-mode automatic
+differentiation](https://github.com/pytorch/pytorch/issues/10223#issuecomment-413935344).
+If one only has one of these implemented, we must perform at least `O(N)`
+evaluations of the function.
+
+### Automatic differentiation in the presence of control flow
+
+One notable absence from all this discussion has been control flow. We
+haven't discussed how gradients flow through control flow, nor mentioned
+control flow at all.
+
+Put simply, control flow can be treated as a static construct. Remember that
+we're taking derivatives, which ask for how our function changes if we add
+epsilon to an input. In that model, our control flow will not change at all,
+and our derivative can be calculated with the exact same operations that we
+performed when evaluating the function.
+
+This is limiting in certain instances, and may not correspond with what we
+believe *should* be the gradient. For example, take
+```
+if m > 0:
+    return x
+else:
+    return 1000 * x
+```
+
+If our goal is optimize the function, then perhaps there should be some kind
+of gradient through `m`. Differentiable relaxations of common control flow
+constructs is an ongoing area of research.
+
+# Notable Challenges
+
+One issue for us is that automatic differentiation is typically applied to
+functions. But, Bril doesn't have functions... :'( What we did was add a
+representation to the Bril JSON format that specified input variables as well
+as initial values for those input variables. As such, much of the usage of
+the auto-differentiation must be done outside of the Bril language itself.
+One thing that would be interesting future work would be to integrate the
+function call work of one the other groups with Bril syntax so that we could
+use auto-differentiation from within Bril.
+
+Another issue that we had was in separating out the different
+auto-differentiation systems. While in other languages, AD systems can often
+be implemented through a combination of custom objects + operator
+overloading, Bril doesn't have either one of these concepts. As such, we
+simply shoved everything inside of the interpreter.
```

Comment:
I'd argue that doing interpretation in the interpreter—even when it's a non-standard interpretation—is actually _more_ elegant than doing it by wonton operator overloading. :smiley:

Reply:
It's more elegant perhaps - but it's also nice to be able to separate autodifferentiation from your core language :)

Reply:
I actually still think operator overloading is nicer - it fits nicer with my model of what's going on. You have your regular data types in your language, and then you have your "derivative enabled" data types.

Reply:
However, fortunately/unfortunately, most languages don't let you overload `for`…